### DISCUSSION TITLE
RFC - New `src` folder structure "module oriented"

### DISCUSSION THREAD

**Discussion by Vadorequest** [UnlyEd/next-right-now#183] [Ideas] [answered]
There had been a few feedbacks regarding the folder structure of the `src` folder.

The goal is to regroup code that is related to each other into modules, represented as an unified folder. Currently, the code is cluttered between "components", "utils" and "types" folders

### Current structure:

![image](https://user-images.githubusercontent.com/3807458/96369708-e1079a00-115b-11eb-9f77-150d6fff1591.png)

### New structure - Proposal 1:

> This structure is theoretical (WIP)

Store everything that is common across pages or too wide to belong into a dedicated module into a `common` folder. Config is also stored into a dedicated folder (instead of `src/`). Then, each module is a folder under `src` and contains all files related to the module. It can, optionally, contain its own `utils`/`types` subfolders if it's too big. 

```
	src
		common
			components
			hocs
			hooks
			stores
			types
			utils
			server
				middlewares
		config
			airtableSchema.ts
			constants.ts
			i18nConfig.js
		pages
			(same as before)
		// modules
		fetchAirtableDataset
			types
			utils
		amplitude
			types
			utils
		api
			types
			utils
		airtableApi
			types
			utils
		vercelCache
			types
			utils
		cookieManager
			types
			utils
		tests
			env.test.ts
			extend-jest
			tests-mocks.ts
			mocks
		sentry
		nextjs
		...
```

### New structure - Proposal 2:

> This structure is theoretical (WIP)

Similar to "**Proposal 1**", but this time everything is under `src/modules` (except for `pages`), which makes it slightly obvious that the approach is module-based.

Similarly, we could have "modules", "common" and "pages" under `src`. It'd be okay too, and could make more sense (let's call it "Proposal 3").

```

	src
		modules
			common
				components
				hocs
				hooks
				stores
				types
				utils
				server
					middlewares
			config
				airtableSchema.ts
				constants.ts
				i18nConfig.js
			fetchAirtableDataset
				types
				utils
			amplitude
				types
				utils
			api
				types
				utils
			airtableApi
				types
				utils
			vercelCache
				types
				utils
			cookieManager
				types
				utils
			tests
				env.test.ts
				extend-jest
				tests-mocks.ts
				mocks
			sentry
			nextjs
			...
		pages
			(same as before)

```

### New structure - Proposal 3:

> This structure is theoretical (WIP)

Similar to "**Proposal 2**", but this time "modules", "common" and "pages" under `src`. It'd be okay too, and could make more sense because "common" isn't really much of a module.

```

	src
		common
			components
			hocs
			hooks
			stores
			types
			utils
			server
				middlewares	
		modules
			config
				airtableSchema.ts
				constants.ts
				i18nConfig.js
			fetchAirtableDataset
				types
				utils
			amplitude
				types
				utils
			api
				types
				utils
			airtableApi
				types
				utils
			vercelCache
				types
				utils
			cookieManager
				types
				utils
			tests
				env.test.ts
				extend-jest
				tests-mocks.ts
				mocks
			sentry
			nextjs
			...
		pages
			(same as before)

```

---

What are your thoughts on this? My personal preference goes for "**Proposal 3**".

**Comment by Vadorequest** [comment]
@jajourda @sovattha @samuelcastro @Demmonius Do you have any preference about this?

**Reply by sovattha** [reply]
Definitely proposal 3 for me. Thanks for asking!

I did not use "modules" much so far, so hiding this complexity away is good for a beginner.

I like the fact that the most important folders are right under src. For me: common, pages, components.

**Reply by Vadorequest** [reply]
> I did not use "modules" much so far, so hiding this complexity away is good for a beginner.

There is no "modules" folder yet, it's a new thing that basically regroup what's currently split between "components", "utils" and "types" (and maybe more).

The thinking it that as soon as you know some code is related and can be grouped together, then you can do so within "modules". Until then, you can do so in "common".

This way, you don't have to think hard when starting something new and just go with "common", and then if at some point it feels right to group together some files then you can create a dedicated module for it.

I believe it'll be easier to maintain this way, because code that relates together will be in a single folder and easier to locate.

**Comment by samuelcastro** [comment]
Definitely **Proposal 3**, this is what I'm using, although I put everything under `common` but I like the `modules` folder.

I proposed something similar before: https://github.com/UnlyEd/next-right-now/pull/44

**Reply by Vadorequest** [reply]
Indeed! Back then (#44) I didn't have a good understanding of what were the gains. Using NRN with an app that grows quite a lot, I now understand better what you meant back then, and I believe NRN can propose both of worlds by proposing common/modules folder for easier understanding of the source code and its growth.

For the record, someone else voted Proposal 3 through direct LinkedIn vote.

I'm not in a hurry to change the structure, so I'll wait a bit to get more feedback before I consider working on it.

**Reply by samuelcastro** [reply]
Are you still planning to move forward with this strategy?

**Reply by Vadorequest** [reply]
Yes, I want to do it. I just lack time and don't know when it'll happen.

**Reply by Vadorequest** [reply]
I'm planning to work on this, this week.

**Reply by samuelcastro** [reply]
Hey @Vadorequest I'm wondering if this is something you're already working? Because since there are a lot of new changes recently, I don't want to merge them and then need to merge a big change of the folder structure again.

If you think this might not get implemented soon, I'm just going to merge the new changes in anyway.

**Reply by Vadorequest** [reply]
On the contrary, I'd rather advise not to do it all at once, it might be too complicated.

You'll probably have a better time merging the current changes now and this later.

And no, I'm not yet working on it. But I'll definitely be doing it this month (🤞 )

**Reply by samuelcastro** [reply]
Ok, thanks for the update.

**Comment by Vadorequest** [comment, marked as answer]
Work has started on https://github.com/UnlyEd/next-right-now/pull/264

@sovattha @samuelcastro Could you review and give your feedback? Does it look like what you expected?

I'm also considering using `@` when loading paths. E.g: `import Button from '@/components/button'`

https://nextjs.org/docs/advanced-features/module-path-aliases

It would be more resilient to moving files around, by not creating conflicts within the file itself. And it'd avoid breaking paths when moving files around (as it happened to me a lot today, had to fix those manually afterwards)

I've never used module path aliasing, do you think it's a good idea to integrate it now? Would it make the upgrade smoother for you?

**Reply by sovattha** [reply]
@Vadorequest Yes, that looks very good. It allows quicker access to the folders likely to be changed: pages, stories, while hiding the complexity of the framework inside a well structured modules or common directories.

I also like the idea to use path aliases. I use them a lot in my projects and it does save time when moving things around.

**Reply by Vadorequest** [reply]
Thanks for your feedback, I played around with module path aliases at https://github.com/UnlyEd/next-right-now/pull/265 and it looks good, also made a codemod so that people can convert all relative links to module aliases in a single step.

**Reply by Vadorequest** [reply]
I spoke too fast, there are compilation issues due to module aliases in Storybook and Jest. I managed to fix storybook, but still haven't found a solution for jest. How did you manage to make it work?

I tried to change the `tsconfig.jest.json` file to:
```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "jsx": "react",
    "baseUrl": ".",
    "paths": {
      "@/common/*": [
        "src/common/*"
      ],
      "@/components/*": [
        "src/common/components/*"
      ],
      "@/utils/*": [
        "src/common/utils/*"
      ],
      "@/modules/*": [
        "src/modules/*"
      ]
    }
  }
}

```

Basically using the same `paths` as the ones used in `tsconfig.json`, I thought that'd work but...

**Reply by Vadorequest** [reply]
Just managed to find a way! I guess it's good for review at https://github.com/UnlyEd/next-right-now/pull/265

let me know if you think some stuff should be done differently about the aliases I defined.

```json
{
      "@/common/*": [
        "src/common/*"
      ],
      "@/components/*": [
        "src/common/components/*"
      ],
      "@/utils/*": [
        "src/common/utils/*"
      ],
      "@/modules/*": [
        "src/modules/*"
      ]
    },
```

**Reply by Vadorequest** [reply]
Also, I was thinking about the basic recommendation about "should I put my file under "common" or make a module"? Here are my 2 cents so far:

One should make a module (`src/modules/`) if:
- At least 2 different entities would be put within the module (types, components, hooks, hocs, utils, ...)
  - If there is only one entity (only components, only utils, etc.) then it probably shouldn't be a module, but belong to `src/common` instead.
- If it is related to an external lib, even if there is only one kind of file (e.g: sentry), then you can make an exception, because having a module for a 3rd party isn't necessarily bad, and helps differentiate our code from code related to 3rd parties
  - It's common with types, because sometimes built-in types aren't the best and we need to overwrite them

What do you think? Do you see other recommendations to share?

**Reply by Vadorequest** [reply]
Updated https://github.com/UnlyEd/next-right-now/pull/265 with support of module aliases for `.js` files too, and used them in Cypress and Storybook as well.

**Reply by Vadorequest** [reply]
@sovattha @samuelcastro Everything looks good to me. I'm awaiting PR review for both #265 and #264. I could merge them right away (they're both stable) but considering how big those changes are I'd very much prefer getting feedback, especially from seasoned developers 😃 

@sovattha Do you use NRN as a boilerplate and keep it up-to-date with the NRN preset too? I know Samuel does.

**Reply by sovattha** [reply]
@Vadorequest My only remark is about this leading slash: I would have named the aliases `@modules`, `@common` instead of `@/modules` or `@/common` but I noticed from the [documentation](https://nextjs.org/docs/advanced-features/module-path-aliases) that it is already like this, so you can leave it as it is.
I used NRN as a boilerplate but I did not keep it up to date because I was happy with the snapshot I started from. Of course, I do not benefit from all the new features (+ I got out of sync with the up to date documentation) and I know that I leave some technical debts open.

**Reply by Vadorequest** [reply]
I believe the form `@/common` is recommended and becoming a standard, because people used to use `common` instead, which is confusing (am I loading an external module or local module? 🤔)

The `@` doesn't work well either since we can have external packages starting with `@` (`@jest/core`, etc.)

That's why the recommended way is `@/`. It makes sense and it's not confusing.

I wrote a small article about the process of migrating a Next.js app from relative paths to module path aliases: https://dev.to/vadorequest/migrating-next-js-jest-storybook-cypress-to-use-module-path-aliases-instead-of-relative-paths-d9a

**Reply by samuelcastro** [reply]
I just see this now, have you fixed the Jest issues you were having?

**Comment by Vadorequest** [comment]
I'm thinking about splitting the block (layout, pageContainer, footer, nav, head) components as some kind of "front-office" module.

My reasoning is that it'd be easier to add a "back-office" module with its own layout, footer, etc. components.

So far, I've considered all those files to be "default" (DefaultLayout, DefaultPageContainer, etc.) but maybe it'd make more sense to modularize them too. Also, it'd make simpler adding another "admin" or "backOffice" module that'd hide behind some kind of authentication, in the future.

I've done something similar with the "NRN examples" (renamed "demo") moved to a "demo" module, with its own components, and "demoSSG/demoSSR" used by "pages/[locale]/demo".

I'm wondering what you think of that, and also what'd be a good default naming. I'd like a name that means "publicly visible part of the site", unlike the "admin" which would be behind auth.

I'm thinking of stuff like:
- modules/siteFront
- modules/siteAdmin

**Reply by sovattha** [reply]
I see what you did in pages/[locale]/demo.
The structure you propose is:
```
index.tsx
|- DefaultLayout
  |- (components used here) -> modules/demo/components
```

If I follow you correctly, by splitting, are you thinking about having a :
1. a `DefaultLayout` and a `AdminLayout`?
2. or a DefaultLayout that would include either **front** and **admin** modules?

Note: The way I implemented authentication: I did not have an admin part as such but part of the website was for authenticated users only. I differentiated authenticated and public components within the `DefaultLayout` directly.
Something like that:
```
index.tsx
|- DefaultLayout
  |- (authenticated components)
  |- (public components)
```
It would have be more elegant indeed to have a separate AdminLayout/AuthenticatedLayout.

Be careful with nesting things in the **modules** folder. Some people might want to use NRN "just" to create simple websites without authentication.
You want to make life easy for them.
Personally, I am afraid to touch anything inside **modules** because it sounds like framework oriented stuff, but I am comfortable to create new pages and components at the root level because I have the feeling I am just adding my stuff on top of something already stable. Yes I would create a `src/components` folder for my own stuff.

What about naming "public" instead of "front"?

**Reply by Vadorequest** [reply]
Thanks for your feedback.

I'll add documentation regarding what modules are and how they can be used in the code itself (using README.md files in those folders (modules, common) directly. Modules aren't meant to be for NRN only, they can totally be extended.

What I'm looking forward to, are:
- Easily add/remove/locate a layout and all its related files
  - Currently, files are split in components, and files related to the `demo` are too much mixed with other files (i.e: `Nav`, etc.)
- One should be able to delete the NRN Demo without affecting the rest of the site, currently it's not too bad but can be better/easier
- Provide a strong foundation/opinion for multiple layouts, if ever needed. And provide a multi-layouts example (default + demo)
- Help share code between pages within the same layout (data fetching, pageProps, etc.)
- Prepare "the field" in case NRN would add an admin layout, or in case people would want to (make it as simple and obvious as possible)
- Maybe build an Admin layout with a single page as example, without authentication, just to show how it can be done

I had time to think about it a bit more since last evening, here's what I've got in mind now:

- common/components/layouts/ will be **removed**
- pages/[locale]/
  - demo/ (pages use the "demo" layout)
  - admin/ (pages use the "admin" layout)
  - privacy, terms pages (pages that are "public")
- modules/app/
  - components/ (AppBootstrap, etc. Types that are shared for the whole app)
  - types/
- modules/appLayouts/
  - demo/
    - components/ 
      - DemoNav, etc. (components related to the demo layout)
    - SSG/SSR utilities (all pages within the same layout need to have common code to fetch data that is shared by all pages within the layout)
  - default/ or public/ or front/ (`public` might be a good choice, `default` too especially if it's being re-used by other layouts and extended)
    - components/
      - DefaultNav or Nav or PublicNav, etc. (components related to this layout)
    - SSG/SSR utilities
  - admin
    - components
      - AdminNav, etc.

This is what I have in mind, it makes it easy to add/remove a layout, and it's intuitive (IMHO) about how things are related.
Also, it helps share business logic and components for the pages within the same layout, and between layouts.

I'm hesitating between "public" and "default" for the base layout, I'd personally prefer "public", and add a "base" layout containing all components and business logic shared between several layouts.

Overall, it complicates things, but that's unfortunately the price for a greater flexibility.

People who want to use only one layout could simply delete the "demo" and "admin" layouts and simply use the public one.
Additionally, I've been thinking about moving the `modules/appLayouts` to `layouts` instead, it'd make it easier to locate and be more obvious.

**Reply by sovattha** [reply]
Providing an example of the `admin` layout without authentication is a good idea, along with `public` (or `default`). I don't intuitively understand the difference between base and default.

About this distinction between pages that need an authentication vs. the ones that do not. I would make it clearer and create 2 sub-directories protected and public, like this:

- pages/[locale]/
  - protected
    - demo/ (pages use the "authenticated demo" layout)
    - admin/ (pages use the "admin" layout)
  - public
    - demo/ (pages use the "un-authenticated demo" layout)
    - privacy, terms pages (pages that are "public")

Even if you consider layouts as a special type of component or module, I would go even a step further for `layouts` and put them at the root level:
- common
- modules
- layouts
- pages
- stories
because in my view, it is a critical piece to modify when we want to customize NRN to our own needs.

You want to make things obvious for new comers and whose who don't read docs.

**Reply by Vadorequest** [reply]
I wouldn't force protected/public folders because those would affect the URL being used by pages, and it's a big tradeoff.

Also, the point is not to differentiate pages being under auth and those public, but to differentiate pages using different layouts. You could use 2 different layouts on your admin site or on your public site. Authentication and layouts are 2 different concepts and while it might seem interesting to merge those two, it'd come at reducing the developer flexibility, and NRN is not meant to do so. (flexibility is one of the most important aspect)

I agree about the layouts location (although I don't call this the "root" folder but the "src" folder). It should be alongside pages, etc.

**Reply by Vadorequest** [reply]
As for "base" vs "default", base is similar to "shared" in this case, it'd be some kind of "meta layout" containing components being used by other layouts. (e.g: You have a Nav component that looks slightly different between your "public" and "admin" layouts, but they share the same base code that is extended in each layout).

I think I'm gonna give it a try with this new structure and see how it feels. I'll let you know for peer-review.

Thank you again for your feedback, it's very useful!

**Reply by sovattha** [reply]
You are welcome, I'll be happy to review. The heavy lifting is all yours. ;)

**Reply by samuelcastro** [reply]
I'd move `module` inside `common`, in my structure for every single page I have a folder under `src` to contain all components/logic that it needs, ie:

- common
  - modules
  - components
  - hooks
  - stores
  - etc
- pages
  - index.tsx
  - search.tsx 
  - admin.tsx
  - etc.tsx
- search
  - components
  - hooks
  - stores
  - etc
  - gql
  - ssg
- home
  - components
  - hooks
  - stores
  - gql
  - ssg
  - etc
- admin
  - components
  - hooks
  - stores
  - gql
  - ssg
  - etc
 
so my `pages/index` and `pages/search` don't have anything other than just import their respective "root" module and SSG/SSR functions like:

```typescript
import { Home, Props, getStaticPropsHomePage } from 'home';
import { getCommonStaticPaths } from '../../common/utils/nextjs/SSG';

export const getStaticPaths: GetStaticPaths<CommonServerSideParams> = getCommonStaticPaths;
export const getStaticProps: GetStaticProps<SSGPageProps, CommonServerSideParams> = getStaticPropsHomePage;

const HomePage: NextPage<Props> = Home;

export default withApollo()(HomePage);
```

Every single component/hook/module/store/etc used by these "root page modules" should live under it and any common code under `common`. This gives me a very concise and clean structure where I easily know where I should put things, and it reduces the conflict between my own work and NRN.

Thoughts?

**Reply by Vadorequest** [reply]
What would `common/modules` contain in your case?

What you're proposing with `home` and `search` sounds like what we call "modules", except they live under the "src/modules" folder.

**Reply by Vadorequest** [reply]
Here is the documented version of what I've done:

https://github.com/UnlyEd/next-right-now/blob/b5adec5b4938a14e31da61986404aa4ef3c1128e/src/README.md

**Reply by samuelcastro** [reply]
`common/modules` contains all NRN native implementation, layouts, utilities, etc.

My root src modules are my own code/modules and they are referenced by a page like:

`./pages/index.tsx` > `./home`
`./pages/search.tsx` > `./search`

**Reply by Vadorequest** [reply]
I see, that's how you avoid running into conflicts with the NRN native code then, I didn't think people would do that. I assumed they'd change the built-in code to adapt it to their need, remove what's not necessary, etc.

**Reply by samuelcastro** [reply]
I still adapt it and remove all unnecessary code, etc, but there are a huge part of it that I just don't need to deeply touch, like i18n, nextjs utilities, SSG bootstrap, etc.

**Reply by Vadorequest** [reply]
I never thought of it that way because I'm so familiar with the NRN base code, it's mine, I own it, and thus I'm not afraid mixing code from NRN and code that is specific to my other projects. But I can see why other people would prefer to split both.

My goal was that forkers would eventually "own" that code, too, and adapt it to their needs and such, but maybe that's "too much code"?

Did you never customize your Sentry/Amplitude configuration? There are many things that can be left untouched, but there are also a few things that I thought people would just change.

**Reply by Vadorequest** [reply]
If splitting code from NRN and your own is important to you, maybe it can be done within the /modules.

Here is how things have been reorganized at this point:
![image](https://user-images.githubusercontent.com/3807458/105077342-aed44400-5a8c-11eb-85f6-375449640e8b.png)

I'm thinking of doing:
- src/modules
  - base (contains all NRN built-in modules, another name is possible)
  - search
  - home
  - admin
  - etc. (basically everything you added yourself)

Another option is what you did, keep "src/modules" for NRN modules and add your own directly into `src/`. 

What do you think about that?

**Reply by samuelcastro** [reply]
I'm using the Sentry/Amplitude config as it is for now by the simple fact that I have many other priorities in the project, considering that my application is still in development, but this is in my backlog.

I'm now well familiar with the NRN code enough to make any changes, however to keep things up to date and be able to easily merge new changes I prefer to organize it this way so that those mergers becomes more manageable.  

And that's probably why I was able to keep up with all those new changes coming almost every day. :)

**Reply by samuelcastro** [reply]
I'm happy with either option (using `modules/myCustomModules` or `src/myCustomModules`) they essentially have the same benefit.  I'd move `layout` to `common` though, unless you have a very specific reason why it should leave under `src`, it's essentially common layouts isn't it?

**Reply by samuelcastro** [reply]
If you're going to keep `app` and `layout` flat on `src` I think having `src/modules/base | home | search` might make more sense. But why not move `app` and `layout` to `common` as well?

**Reply by Vadorequest** [reply]
> I'm happy with either option (using modules/myCustomModules or src/myCustomModules) they essentially have the same benefit. 

I see, I think I'll keep it that way then, under src/modules for NRN built-in modules, and people can do it either way to add their own modules (/src/modules or /src) depending on their personal preference.

> I'd move layout to common though, unless you have a very specific reason why it should leave under src, it's essentially common layouts isn't it?

I see what you mean, but "layouts" are a kind-of modules we wished to highlight instead of burying them under src/modules. Also, they use a very different design pattern that doesn't fit with the "common" folder (common = MVC-ish) since each layout can hold its own components, business logic (data fetching, types, hooks), etc. I think it makes more sense to have them in src/ directly, they're more discoverable and they fit better this way IMHO (modular approach).

**Reply by samuelcastro** [reply]
Will you keep all NRN under `src/modules/base` then? If you do that I'm probably going to switch to this approach as well and add all my custom modules under `src/modules` as well, essentially what you described before:

- src/modules
  - base (contains all NRN built-in modules, another name is possible)
  - search
  - home
  - admin
  - demo (you could add the demo here as well)
  - etc

**Reply by Vadorequest** [reply]
I could do that, I'm not 100% sure it's the right decision or naming ("base"), but I'm inclined to split NRN modules and custom modules considering the clarity it brings for forkers.

The "demo" being a layout, essentially composed of components and a few files for data fetching, I feel better having all those grouped together under /src/layouts/demo, it makes more sense to have all the pieces there.

In the end, there is no one-size-fit-all way of doing things, the goal is to find something as good as possible for as many people/projects as possible.

**Reply by samuelcastro** [reply]
> In the end, there is no one-size-fit-all way of doing things, the goal is to find something as good as possible for as many people/projects as possible.

That make sense.

**Reply by samuelcastro** [reply]
So after think more about it, I'll keep my custom modules/work under `src` close to `layout` it'll make more sense since they will share similar data structure (components, hooks, stores, etc).

**Reply by Vadorequest** [reply]
Okay, on my side I'm thinking about `core`, or `lib` instead of `base`. I'll keep thinking a little while about it;

**Reply by Vadorequest** [reply]
@samuelcastro  @sovattha I think I'm done. Please review :)

Here's the branch, easier to "feel" the new folder structure:
https://github.com/UnlyEd/next-right-now/tree/folder-restructure/src

PR: https://github.com/UnlyEd/next-right-now/pull/264 (very slow due to almost 400 files changed)

I'll sleep on it and merge it tomorrow if it looks good for you.

It'll be hard to merge into our private forks, I expect hundreds of conflicts. 😭  But it'll be much better afterwards. (and I don't plan on changing the folder structure ever again 😅 )

**Reply by sovattha** [reply]
@samuelcastro @Vadorequest I just kept up with your conversation.

Indeed, I think `core` is a good name, instead of `base`.

I like the fact that all the custom code could live flat at the `src` level.

I reviewed the changes as best as I could, but 400 files is a challenge!

**Reply by Vadorequest** [reply]
> Indeed, I think core is a good name, instead of base.

👍 

> I like the fact that all the custom code could live flat at the src level.

I also like the fact it's flexible enough for people to choose between /src or /src/modules at their convenience.
I think it might not be such a good design for huge projects with 20-ish modules living alongside /pages, but for small to middle-sized projects it would fit well. People can always move them to /src/modules if they wish so.

Would you keep NRN built-in modules in src/modules/core then? Do you like having them separated to locate your own code better?

> I reviewed the changes as best as I could, but 400 files is a challenge!

Nobody can review that much, I'm only confident in this update because of CI/CD testing everything every time I pushed stuff, it'd be a nightmare without such automation.

**Reply by sovattha** [reply]
> Would you keep NRN built-in modules in src/modules/core then? Do you like having them separated to locate your own code better?

Yes I would do so. I like to have at hand the folders/files that I am supposed to edit as a newbie along with the ability to use modules when the project reaches a critical size.

**Reply by Vadorequest** [reply]
Alright! I'll let things as they are now then, thank you both for your valuable feedback!

I'll merge today, once I've merged don't rush upgrading, I'll need to do the same work on the other preset, and then I'll also need to adapt my app and I'll share best practices and feedback experience about upgrading then.

**Reply by samuelcastro** [reply]
I just finished the review, everything looks great, I'm really excited about it, I was looking forward to see this change. 😄 

The next big DX improvement I'd do is to include Prettier, I know we discussed this [before](https://github.com/UnlyEd/next-right-now/pull/76), but without Prettier it's really hard to keep a consistent code style across different devs, but anyway it's a discussion for another time I guess. =)

Thanks for this hard work @Vadorequest. 

PS: I also prefer `core`.

**Comment by Vadorequest** [comment]
@samuelcastro I migrated the `gcms` preset to the new folder structure. See PR https://github.com/UnlyEd/next-right-now/pull/275

Everything looks good to me.

I found a few things I had missed, which were fixed directly in the above PR for `gcms`, they'll be fixed in a [separate PR for the `at`](https://github.com/UnlyEd/next-right-now/pull/274) preset.

**Reply by Vadorequest** [reply]
Just found quite a few issues 😬  still on it.

**Reply by Vadorequest** [reply]
I merged the #275 PR, but still advise to wait a bit before migrating.

**Reply by samuelcastro** [reply]
Ok, what are your concerns?

And what are those "quite few issues"?

**Reply by samuelcastro** [reply]
I think you forgot to move the `gql` folder, see: https://github.com/UnlyEd/next-right-now/tree/v2-mst-aptd-gcms-lcz-sty/src

**Reply by Vadorequest** [reply]
> And what are those "quite few issues"?

Path issues, etc. nothing that couldn't be fixed :)

I just wanted to be the first to go through the pain of migrating, it's done, it took me about 3h.

> I think you forgot to move the gql folder, see: v2-mst-aptd-gcms-lcz-sty/src

Actually, I thought about leaving it there, didn't really gave much thoughts to be honest. But we could move/split it if you think that's better. I like how it reproduces the pages structure, it's not a bad place for it IMHO.

**Reply by samuelcastro** [reply]
I'd move it to `common` I think it makes more sense.

**Reply by Vadorequest** [reply]
Depends, if you have one module per page (I kinda do) then it'd make more sense to have each gql query within the module.

**Reply by Vadorequest** [reply]
But I don't really care to be honest, as I don't use that preset myself anymore :D

**Reply by samuelcastro** [reply]
> Depends, if you have one module per page (I kinda do) then it'd make more sense to have each gql query within the module.

That's what I do, each module has its own `gql` and all common queries I leave in my `common/gql`.

**Reply by Vadorequest** [reply]
Yeah I get it now, makes total sense.

**Reply by Vadorequest** [reply]
I'll do it now.

**Reply by Vadorequest** [reply]
Done :)

**Reply by samuelcastro** [reply]
Awesome!

**Comment by Vadorequest** [comment]
Okay, so I managed to merge all those changes into our private NRN fork.

We got about 100-150 conflicts, lots of things can be easily solved because they're about /demo. (about half of the conflicts)
The whole process took about 3h. I advise to do it if you can have 3h free of disturbance.

- Merge changes
- Fix conflicts (30mn)
- Once I fixed all conflicts, the merge operation broke and I had to `git rm` a particular file to continue the merge 🤷 
- Fix all broken paths (1-2h) this was the most painful part, I'd suggest creating the directory and moving the core files in advance so that paths can be easily replaced instead of doing it manually after the merge, but I haven't tried that way, I went for the long and painful way of fixing all paths by hand
- Fix regressions (SSG/SSR, multiversalAppBootstrap and shared components (Nav, Footer, etc.) in particular (20mn)
- Move my own code to the right spot (10mn)
- Run codemod `yarn codemod:module-path-aliases` (do that only if you've committed everything) but the codemod broke because of some babel issue, it was working fine until now 🤷 

Had about 500 modified files in total.

Good luck 😄
